presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -314,18 +315,50 @@ public FileStatus getFileStatus(Path path)
         }
 
         return new FileStatus(
-                getObjectSize(metadata),
+                getObjectSize(metadata, path),
                 false,
                 1,
                 BLOCK_SIZE.toBytes(),
                 lastModifiedTime(metadata),
                 qualifiedPath(path));
     }
 
-    private static long getObjectSize(ObjectMetadata metadata)
+    private long getObjectSize(ObjectMetadata metadata, Path path) throws IOException
     {
         String length = metadata.getUserMetadata().get(UNENCRYPTED_CONTENT_LENGTH);
-        return (length != null) ? Long.parseLong(length) : metadata.getContentLength();
+
+        if (length != null) {
+            return Long.parseLong(length);
+        }
+
+        // check if file has KMS encryption
+        String cryptoAlg = metadata.getUserMetadata().get(CRYPTO_KEYWRAP_ALGORITHM);
+        if (cryptoAlg != null && cryptoAlg.equalsIgnoreCase("kms")) {
+            // KMS client-side encryption
+            long decryptedLength = metadata.getContentLength();
+
+            // load file and determine actual length
+            FSDataInputStream inputStream = this.open(path);
+
+            // read last 25 bytes of file to help determine length
+            int eofOffset = 25;
+            long startingPos = (decryptedLength > eofOffset) ? decryptedLength - eofOffset : 0;
+
+            // seek to starting point
+            inputStream.seek(startingPos);
+
+            while (inputStream.read() != -1) {
+                startingPos++;
+            }
+            decryptedLength = startingPos;
+
+            inputStream.close();
+            return decryptedLength;
```

Comment:
you can just return `startingPos` instead of updating and returning `decryptedLength`.